File: 프로그래머스/3/49191. 순위/순위.java
class Solution {
    public int solution(int n, int[][] results) {
        int answer = 0;
        
        // 2차원 인접행렬
        int[][] graph = new int[n][n];
        
        // 이기면 1 지면 -1
        for (int[] result : results) {
            int a = result[0], b = result[1];
            graph[a-1][b-1] = 1;
            graph[b-1][a-1] = -1;
        }
        
        // 플로이드 와샬
        for (int k = 0; k < n; k ++) {
            for (int i = 0; i < n; i ++) {
                for (int j = 0; j < n; j++) {
                    if (i == j || graph[i][j] != 0)
                        continue;
                    if (graph[i][k] == 1 && graph[k][j] == 1) {
                        graph[i][j] = 1;
                        graph[j][k] = -1;
                        graph[k][i] = -1;
                        graph[j][i] = -1;
                    }
                }
            }
        }
        
        // 0이 하나인 애들은 확정
        for (int i = 0; i<n ;i++) {
            int cnt = 0;
            
            for (int j = 0; j < n; j++) {
                if (graph[i][j] == 0)
                    cnt += 1;
            }
            if (cnt == 1)
                answer++;
        }
        
        
        return answer;
    }
}

File: 프로그래머스/3/49191. 순위/순위.py
# 

def solution(n, results):
    answer = 0
    
    board = [[0] * (n+1) for _ in range(n+1)]
    
    for win, lose in results:
        board[win][lose] = 1
        board[lose][win] = -1
    
    for k in range(1, n+1):
        for i in range(1, n+1):
            for j in range(1, n+1):
                if i == j or board[i][j] != 0:
                    continue
                else:
                    if board[i][k] != 0 and board[i][k] == board[k][j] :
                        board[i][j] = board[i][k]
                        board[j][i] = board[i][k] * -1
    
    for result in board:
        if result.count(0) == 2:
            answer+=1
    return answer